### PR TITLE
topk: allow negative axis

### PR DESCRIPTION
**Pull request by josel-amd**
Matches the onnx definition

### CODE REVIEW

File: lib/Dialect/XTenNN/IR/XTenNNOps.cpp [hunk 1]
```diff
@@ -552,11 +552,21 @@ LogicalResult TopK::inferReturnTypeComponents(
...
 
-  auto axis = adaptor.getAxis();
-  if (axis >= (uint64_t)inTy.getRank()) {
-    return emitOptionalError(location, "expected axis <= rank of input");
+  auto axis = (int64_t)adaptor.getAxis();
+  // onnx spec: axis: [-r, r-1]
+  if (!(axis >= -inTy.getRank()) || !(axis < inTy.getRank())) {
+    return emitOptionalError(
+        location,
+        "expected axis to be within \"rank < axis <= rank - 1\" of input");
```
Comment (cferry-AMD) [inline lib/Dialect/XTenNN/IR/XTenNNOps.cpp]:
```suggestion
        "expected axis to be within \"-rank <= axis < rank\" of input");
```

File: test/Dialect/XTenNN/ops_invalid.mlir [hunk 2]
```diff
@@ -93,7 +93,7 @@ func.func @topk_wrong_indices_shape(%arg0: tensor<10x10xf32>) {
 func.func @topk_wrong_axis(%arg0: tensor<10x10xf32>) {
     %k = arith.constant 7 : i64
     // expected-error@+2 {{failed to infer returned types}}
-    // expected-error@+1 {{expected axis <= rank of input}}
+    // expected-error@+1 {{expected axis to be within "rank < axis <= rank - 1" of input}}
```
Comment (cferry-AMD) [inline test/Dialect/XTenNN/ops_invalid.mlir]:
```suggestion
    // expected-error@+1 {{expected axis to be within "-rank <= axis < rank" of input}}
```

File: test/Dialect/XTenNN/ops_invalid.mlir [hunk 3]
```diff
@@ -107,3 +107,13 @@ func.func @topk_large_k(%arg0: tensor<10x10xf32>) {
     %a, %b = xten_nn.topk(%arg0 : tensor<10x10xf32>, %k : i64) {axis = 0 : i64, largest = true, sorted = true} -> tensor<10x10xf32>, tensor<1xi64>
     return
 }
+
+// -----
+
+func.func @topk_negative_axis(%arg0: tensor<10x10xf32>) {
+    %k = arith.constant 100 : i64
+    // expected-error@+2 {{failed to infer returned types}}
+    // expected-error@+1 {{expected axis to be within "rank < axis <= rank - 1" of input}}
```
Comment (cferry-AMD) [inline test/Dialect/XTenNN/ops_invalid.mlir]:
```suggestion
    // expected-error@+1 {{expected axis to be within "-rank <= axis < rank" of input}}
```

**Review by cferry-AMD — APPROVED** [review]
Thanks!

File: lib/Dialect/XTenNN/IR/XTenNNOps.cpp [hunk 4]
```diff
@@ -552,11 +552,21 @@ LogicalResult TopK::inferReturnTypeComponents(
 
   auto inTy = cast<RankedTensorType>(adaptor.getInput().getType());
 
-  auto axis = adaptor.getAxis();
-  if (axis >= (uint64_t)inTy.getRank()) {
-    return emitOptionalError(location, "expected axis <= rank of input");
+  auto axis = (int64_t)adaptor.getAxis();
+  // onnx spec: axis: [-r, r-1]
+  if (!(axis >= -inTy.getRank()) || !(axis < inTy.getRank())) {
```
Comment (TinaAMD) [inline lib/Dialect/XTenNN/IR/XTenNNOps.cpp]: Nit for readability:
```suggestion
  if (axis < -inTy.getRank() || axis => inTy.getRank())) {
```

File: lib/Dialect/XTenNN/IR/XTenNNOps.cpp (hunk 1 above)
Comment (TinaAMD) [reply lib/Dialect/XTenNN/IR/XTenNNOps.cpp]: Maybe
> expected axis to be within `[-rank, rank)` (where `rank` is the rank of the input)
(or `[-rank, rank-1]`)

File: lib/Dialect/XTenNN/IR/XTenNNOps.cpp [hunk 5]
```diff
@@ -552,11 +552,21 @@ LogicalResult TopK::inferReturnTypeComponents(
...
+        location,
+        "expected axis to be within \"rank < axis <= rank - 1\" of input");
   }
 
+  // normalize axis: [0, r)
+  if (axis < 0) {
+    axis += inTy.getRank();
+  }
+
+  assert((axis >= 0 && axis < inTy.getRank()) && "axis with wrong value");
```
Comment (TinaAMD) [inline lib/Dialect/XTenNN/IR/XTenNNOps.cpp]:
```suggestion
  assert((axis >= 0 && axis < inTy.getRank()) && "axis has invalid value");
```

**Review by TinaAMD — APPROVED** [review]
LGTM

File: lib/Dialect/XTenNN/IR/XTenNNOps.cpp (hunk 1 above)
Comment (josel-amd) [reply lib/Dialect/XTenNN/IR/XTenNNOps.cpp]: Fixed

**Pull request merged**